_posts/2021-04-22-torch-for-optimization/torch-for-optimization.Rmd

```diff
@@ -0,0 +1,640 @@
+
+Specifically, `torch`'s L-BFGS optimizer implements the [Strong Wolfe conditions](https://en.wikipedia.org/wiki/Wolfe_conditions). We re-run the above code, changing just two lines. Most importantly, the one where the optimizer is instantiated:
+
+```{r}
+optimizer <- optim_lbfgs(x_star, line_search_fn = "strong_wolfe")
+```
+
+And secondly, this time I found that after the third iteration, loss continued to decrease for a while, so I let it run for five iterations. Here is the output:
+
+    Iteration:  1 
+    ...
+    ...
+
+    Value is:  -0.8838741 
+    Gradient is:  3.742207 7.521572 
+    X is:  0.09035123 -0.03220009 
+
+    Value is:  -0.928809 
+    Gradient is:  1.464702 0.9466625 
+    X is:  0.06564617 -0.026706 
+
+    Iteration:  2 
+    ...
+    ...
+
+    Value is:  -0.9991404 
+    Gradient is:  39.28394 93.40318 
+    X is:  0.0006493925 -0.0002656128 
+
+    Value is:  -0.9992246 
+    Gradient is:  6.372203 12.79636 
+    X is:  0.0007130796 -0.0002947929 
+
+    Iteration:  3 
+    ...
+    ...
+
+    Value is:  -0.9997789 
+    Gradient is:  3.565234 5.995832 
+    X is:  0.0002042478 -8.457939e-05 
+
+    Value is:  -0.9998025 
+    Gradient is:  -4.614189 -13.74602 
+    X is:  0.0001822711 -7.553725e-05 
+
+    Iteration:  4 
+    ...
+    ...
+
+    Value is:  -0.9999917 
+    Gradient is:  -382.3041 -921.4625 
+    X is:  -6.320081e-06 2.614706e-06 
+
+    Value is:  -0.9999923 
+    Gradient is:  -134.0946 -321.2681 
+    X is:  -6.921942e-06 2.865841e-06 
+
+    Iteration:  5 
+    ...
+    ...
+
+    Value is:  -0.9999999 
+    Gradient is:  -3446.911 -8320.007 
+    X is:  -7.267168e-08 3.009783e-08 
+
+    Value is:  -0.9999999 
+    Gradient is:  -3419.361 -8253.501 
+    X is:  -7.404627e-08 3.066708e-08 
+
+It's still not perfect, but a lot better.
+
+Finally, let's go one step further. Can we use `torch` for constrained optimization?
+
+### Quadratic penalty for constrained optimization
+
+In constrained optimization, we still search for a minimum, but that minimum can't reside just anywhere: Its location has to fulfill some number of additional conditions. In optimization lingo, it has to be *feasible*.
+
+To illustrate, we stay with the flower function, but add on a constraint: $\mathbf{x}$ has to lie outside a circle of radius $sqrt(2)$, centered at the origin. Formally, this yields the inequality constraint
+
+$$
+2 - {x_1}^2 - {x_2}^2 <= 0
+$$
+
+A way to minimize *flower* and yet, at the same time, honor the constraint is to use a penalty function. With penalty methods, the value to be minimized is a sum of two things: the target function's output and a penalty reflecting potential constraint violation. Use of a *quadratic* *penalty*, for example, results in adding a multiple of the square of the constraint function's output:
```

Comment:
Why is flower italicized in the first sentence here?

Reply:
> A way to minimize *flower* and yet, at the same time, honor the constraint is to use a penalty function

Maybe simplify this sentence a bit?
“One way to minimize flower while still honoring the constraint is to use a penalty function."

Reply:
italicized because you don't normally normalize flowers :-)))
(meaning, otherwise I would have felt I need the full "flower function" yet again :-)

About the simplification, not sure... I don't really find it complicated and of course, like its more "oral" sound